src/main/java/org/spongepowered/api/profile/GameProfileCache.java

```diff
@@ -0,0 +1,313 @@
+     * @return {@code true} if the profile was successfully cached,
+     * otherwise {@code false}
+     */
+    default boolean add(GameProfile profile) {
+        return this.add(profile, null);
+    }
+
+    /**
+     * Add an entry to this cache, with an optional expiration date.
+     *
+     * @param profile The profile to cache
+     * @param expiry The expiration date
+     * @return {@code true} if the profile was successfully cached,
+     * otherwise {@code false}
+     */
+    default boolean add(GameProfile profile, @Nullable Date expiry) {
+        return this.add(profile, false, expiry);
+    }
+
+    /**
+     * Add an entry to this cache, with an optional expiration date.
+     *
+     * @param profile The profile to cache
+     * @param overwrite If we should overwrite the cache entry for
+     * the provided profile
+     * @param expiry The expiration date
+     * @return {@code true} if the profile was successfully cached,
+     * otherwise {@code false}
+     */
+    boolean add(GameProfile profile, boolean overwrite, @Nullable Date expiry);
+
+    /**
+     * Gets a {@link GameProfile} from this cache by its unique id.
+     *
+     * @param uniqueId The unique id of the profile
+     * @return The profile, if present, or {@link Optional#empty()} if
+     * the cache did not contain a profile with the provided id
+     */
+    default Optional<GameProfile> get(UUID uniqueId) {
+        checkNotNull(uniqueId, "unique id");
+        Collection<GameProfile> profiles = this.getBulkId(Collections.singleton(uniqueId));
+        if (profiles.isEmpty()) {
+            return Optional.empty();
+        }
+
+        return Optional.ofNullable(Iterables.getFirst(profiles, null));
+    }
+
+    /**
+     * Gets {@link GameProfile}s in bulk by their unique id.
+     *
+     * @param uniqueIds The unique ids
+     * @return A collection of successfully found up profiles
+     */
+    Collection<GameProfile> getBulkId(Iterable<UUID> uniqueIds);
+
+    /**
+     * Looks a {@link GameProfile} up by its unique id and
+     * loads it into this cache.
+     *
+     * @param uniqueId The unique id of the profile
+     * @return The profile, if present, or {@link Optional#empty()} if
+     * we couldn't find a profile with the provided id
+     */
+    default Optional<GameProfile> lookup(UUID uniqueId) {
+        checkNotNull(uniqueId, "unique id");
+        Collection<GameProfile> profiles = this.lookupBulkId(Collections.singleton(uniqueId));
+        if (profiles.isEmpty()) {
+            return Optional.empty();
+        }
+
+        return Optional.ofNullable(Iterables.getFirst(profiles, null));
+    }
+
+    /**
+     * Looks up {@link GameProfile}s in bulk by their unique id and
+     * loads them into this cache.
+     *
+     * @param uniqueIds The unique ids
+     * @return A collection of successfully looked up profiles
+     */
+    Collection<GameProfile> lookupBulkId(Iterable<UUID> uniqueIds);
```

Comment:
* Single `lookupId`
* Bulk `lookupIds`

I would also default implement the bulk methods using the single method, because some impls aren't capeable to do real bulk lookups anyway. If an impl can optimize performance by using custom impls they can still overwrite the impl if they want.